src/utils/derive_offset_generators.nr

```diff
@@ -492,30 +495,162 @@ where
+        // compute_and_print_offset_generators::<Secp384r1BN, Secp384r1PartialCurveParams, _, _, 97>("Secp384r1_Fq", "Secp384r1_Params", 1);
         // cofactor = 0x170b5d44300000000000000000000000 booooo
     }
 }
+
+mod tests {
+    use bignum::BigNum;
+    use bignum::BLS12_377_Fq;
+    use bignum::BLS12_381_Fq;
+    use bignum::BN254_Fq;
+    use bignum::MNT4_753_Fq;
+    use bignum::MNT6_753_Fq;
+    use bignum::Pallas_Fq;
+    use bignum::Secp256k1_Fr;
+    use bignum::Secp256r1_Fr;
+    use bignum::Secp384r1_Fr;
+    use bignum::Vesta_Fq;
+
+    use crate::{BigCurve, BigCurveTrait, curve_jac::CurveJ, CurveParamsTrait};
+
+    use super::BLS377PartialCurveParams;
+    use super::BLS381PartialCurveParams;
+    use super::BN254PartialCurveParams;
+    use super::Mnt4PartialCurveParams;
+    use super::Mnt6PartialCurveParams;
+    use super::PallasPartialCurveParams;
+    use super::Secp256k1PartialCurveParams;
+    use super::Secp256r1PartialCurveParams;
+    use super::Secp384r1PartialCurveParams;
+    use super::VestaPartialCurveParams;
+
+    unconstrained fn verify_offset_generators<Fq, Curve>(cofactor: Field)
+    where
+        Fq: BigNum,
+        Curve: CurveParamsTrait<Fq>,
+    {
+        // For each curve, compute the offset generators and verify they match the ones in the curve parameters
+        let seed: [u8; 25] = "bigcurve offset generator".as_bytes();
+
+        // we only need well-defined parameters for a and b, for hash_to_curve to work
+        // (no offset generators needed)
+        let g: BigCurve<Fq, Curve> = BigCurve::hash_to_curve(seed);
+
+        let g_init: CurveJ<Fq, Curve> = CurveJ { x: g.x, y: g.y, z: Fq::one(), is_infinity: false };
+        let mut gen = g_init;
+
+        let cofactor_bits: [u1; 128] = cofactor.to_be_bits();
+        let mut it: u32 = 0;
+        for i in 0..128 {
+            if (cofactor_bits[i] == 1) {
+                it = i + 1;
+                break;
+            }
+        }
+
+        for i in it..128 {
+            gen = gen.dbl().0;
+            if (cofactor_bits[i] == 1) {
+                gen = gen.incomplete_add(g_init).0;
+            }
+        }
+        let mut genz = gen.z;
+        genz = genz.__invmod();
+        let zz = genz.__mul(genz);
+        let zzz = zz.__mul(genz);
+        let computed_initial = [gen.x.__mul(zz), gen.y.__mul(zzz)];
+        // Verify against the values in the curve parameters
+        let expected_initial = Curve::offset_generator();
+        println(f"expected_initial: {expected_initial}");
+        println(f"computed_initial: {computed_initial}");
+        assert_eq(computed_initial[0], expected_initial[0], "Initial offset generator x mismatch");
+        assert_eq(computed_initial[1], expected_initial[1], "Initial offset generator y mismatch");
+    }
+
+    #[test]
+    unconstrained fn verify_bn254_offset_generators() {
+        // BN254 cofactor is 1
```

Comment:
Not a blocker for this PR but it seems like we should add the cofactor onto the `CurveParamsTrait` trait.